external/kafka/src/main/scala/org/apache/spark/rdd/kafka/OffsetRange.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rdd.kafka
+
+/** Represents a range of offsets from a single Kafka TopicAndPartition */
+trait OffsetRange {
```

Comment:
No, it's not attaching the offset to every record, that's what i'm trying
to avoid.  It's dealing with the offsets either once per rdd, or once per
partition, depending on what is necessary from a client semantics point of
view.

Hopefully accessing the RDD from inside foreachRDD isn't contentious?

As for accessing the partition, yeah its pretty weird to have to go index
the rdd to get the partition... but it's also pretty weird that the
existing apis named 'mapPartition*' don't actually give you access to the
partition... after all, the partition is serializable.



On Wed, Jan 28, 2015 at 11:47 PM, Tathagata Das <notifications@github.com>
wrote:

> In
> external/kafka/src/main/scala/org/apache/spark/rdd/kafka/OffsetRange.scala
> <https://github.com/apache/spark/pull/3798#discussion_r23747942>:
>
> > + * (the "License"); you may not use this file except in compliance with
> > + * the License.  You may obtain a copy of the License at
> > + *
> > + *    http://www.apache.org/licenses/LICENSE-2.0
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.rdd.kafka
> > +
> > +/** Represents a range of offsets from a single Kafka TopicAndPartition */
> > +trait OffsetRange {
>
> Right, I get your point. Though I thought about the usage based on the
> example, and I think we need to think this a little bit more. From what I
> understood, you are attaching the offset in every records, and shuffling
> everything with that offset attached. That is quite a loss of efficiency.
> Also, accessing the RDD and its partition object from within the
> mapPartition function is very confusing, and ... does it actually work????
> If at all this works, thats not even the recommended RDD operation!
>
> We really need to come up with a better way to expose offsets.
> Brainstorming a little more on this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3798/files#r23747942>.
>